src/PIL/Jpeg2KImagePlugin.py

```diff
@@ -114,11 +178,10 @@ def _parse_jp2_header(fp):
                     mode = "RGB"
                 elif nc == 4:
                     mode = "RGBA"
-                break
         elif tbox == b"colr":
-            meth, prec, approx = struct.unpack_from(">BBB", content)
-            if meth == 1:
-                cs = struct.unpack_from(">I", content, 3)[0]
+            meth, prec, approx = header.read_fields(">BBB")
+            if meth == 1 and unkc == 0:
```

Comment:
https://www.ics.uci.edu/~dan/class/267/papers/jpeg2000.pdf
> Colourspace Unknown. This field specifies if the actual colourspace of the image data is known. This field is encoded as a 1-byte unsigned integer. Legal values for this field are 0, if the colourspace of the image is known and correctly specified the colourspace boxes within the file, or 1, if the colourspace of the image is not known. A value of 1 will be used in cases such as the transcoding of legacy images where the actual colourspace of the image data is not known. In those cases, while the colourspace interpretation methods specified in the file may not accurately reproduce the image with respect to some original, the image should be treated as if the methods do accurately reproduce the image.

The way I read that, even if `unkc` is 1, we should still use `colr` regardless, so this change isn't needed. Any thoughts?

Reply:
The reason for this change was to preserve the previous behavior of the implementation: If an "ihdr" box was found with unkc != 0, we'd break from the loop without reading from "colr": https://github.com/python-pillow/Pillow/blob/3346389ace0537b17360928ca59bff0a50f94041/src/PIL/Jpeg2KImagePlugin.py#L103-L117

---

That being said, I agree with your reading of the spec. As I interpret it: while the colorspace information may be inaccurate, we are to treat it as if it were accurate, and thus read it from "colr".

However, if this is the case, then we may as well stop trying to infer `mode` from the number of channels (`nc` in "ihdr") if `unkc` is set, because then we should only assign `mode` based on the values in the "colr" header.

Unless... would we want to accept unsupported `cs` / `nc` values in the "colr" header if `unkc` is set? That's the behavior you'd get if you'd just remove `and unkc == 0` from the commented-on if statement. I'm really unsure about the correct behavior in such a situation.

Nevertheless, I think that a change that alters how the colorspace information is parsed is outside the scope of this PR 😄

Reply:
Thanks for the explanation. Yep, fair enough trying to keep the behaviour the same.